ipq40xx: add support for P&W R619AC (aka G-DOCK 2.0)

ysc3839 opened the pull request:
ipq40xx: add support for P&W R619AC (aka G-DOCK 2.0)

P&W R619AC is a IPQ4019 Dual-Band AC1200 router.
It is made by P&W (p2w-tech.com) known as P&W R619AC but marketed and sold more popularly as G-DOCK 2.0.

Specification:

* SOC: Qualcomm Atheros IPQ4019 (717 MHz)
* RAM: 512 MiB
* Flash: 16 MiB (NOR) + 128 MiB (NAND)
* Ethernet: 5 x 10/100/1000 (4 x LAN, 1 x WAN)
* Wireless:
  - 2.4 GHz b/g/n Qualcomm Atheros IPQ4019
  - 5 GHz a/n/ac Qualcomm Atheros IPQ4019
* USB: 1 x USB 3.0
* LED: 4 x LAN, 1 x WAN, 2 x WiFi, 1 x Power (All Blue LED)
* Input: 1 x reset
* 1 x MicroSD card slot
* Serial console: 115200bps, pinheader J2 on PCB
* Power: DC 12V 2A
* 1 x Unpopulated mPCIe Slot
* 1 x Unpopulated Sim Card Slot

Installation:
1. Access to tty console via UART serial
2. Enter failsafe mode and mount rootfs (https://openwrt.org/docs/guide-user/troubleshooting/failsafe_and_factory_reset)
3. Edit inittab to enable shell on tty console `sed -i 's/#ttyM/ttyM/' /etc/inittab`
4. Reboot and upload `-nand-factory.bin` to the router (using wget)
5. Use `sysupgrade` command to install

Another installation method is to hijack the upgrade server domain of stock firmware, because it's using insecure http.

This commit is based on @LGA1150(at GitHub)'s work https://github.com/LGA1150/openwrt/commit/a4932c8d5a275d1fb4297bd20ec03f9270a45d1c
With some changes:
1. Added `qpic_bam` node in dts. I don't know much about this, but I observed other dtses have this node.
2. Removed `ldo` node under `sd_0_pinmux`, because `ldo` cause SD card not working. This fix is from https://github.com/coolsnowwolf/lede/commit/51143b4c7571f717afe071db60bbb4db1532cbf2
3. Removed the 32MB NOR variant.
4. Removed `cd-gpios` in `sdhci` node, because it's reported that it makes wlan2g led light up.
5. Added ethphy led config in dts.

About the 128MiB variant: The stock bootloader hardcoded 64MiB size of nand. But most of this devices have 128MiB nand. If you want to use all 128MiB, you need a modified bootloader.

Signed-off-by: Richard Yu <yurichard3839@gmail.com>

ysc3839 commented:
Updated installation description.

ysc3839 commented:
USB 3.0 and MicroSD is working. I think it's ready for review.
```
root@OpenWrt:~# lsusb -t
/:  Bus 04.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/0p, 5000M
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 480M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 5000M
    |__ Port 1: Dev 2, If 0, Class=Mass Storage, Driver=usb-storage, 5000M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci-hcd/1p, 480M
```
```
root@OpenWrt:~# dd if=/dev/sda2 of=/dev/null bs=8k status=progress
1207754752 bytes (1.2 GB, 1.1 GiB) copied, 6 s, 201 MB/s

root@OpenWrt:~# dd if=/dev/zero of=/dev/sda2 bs=4k status=progress
2838982656 bytes (2.8 GB, 2.6 GiB) copied, 20 s, 142 MB/s
```
```
root@OpenWrt:~# dd if=/dev/mmcblk0 of=/dev/null bs=4k status=progress
1144709120 bytes (1.1 GB, 1.1 GiB) copied, 17 s, 67.3 MB/s

root@OpenWrt:~# dd if=/dev/zero of=/dev/mmcblk0 bs=4k status=progress
144482304 bytes (144 MB, 138 MiB) copied, 7 s, 20.6 MB/s
```

AmadeusGhost commented:
```
define Device/p2w_r619ac
	IMAGES += nand-factory.bin
	IMAGE/nand-factory.bin := append-ubi | qsdk-ipq-factory-nand
endef
TARGET_DEVICES += p2w_r619ac

define Device/p2w_r619ac-128m
	$(call Device/p2w_r619ac)
	DEVICE_VARIANT := 128M NAND
endef
TARGET_DEVICES += p2w_r619ac-128m
```
This nand-factory.bin only works in stock firmware, 128m version does not need that.
```
&sdhci {
	cd-gpios = <&tlmm 22 GPIO_ACTIVE_LOW>;
};
```
I personally suggest to remove this "cd-gpios" line, which can make the wlan2g led resume to light up after the SD card is pulled out. (Due to the gpio of sd and wlan2g led are multiplexed.)
> WAN and LAN LEDs are broken, possible a driver bug: coolsnowwolf/lede@9135c94

It needs this patch to work: https://github.com/x-wrt/x-wrt/commit/7333152f8e5b6ee72abac3ffb88d7a4ee29b2872 . `ar40xx_phy_mmd_write(priv, 0x1f, 7, 0x8076, 0x0670); /* 1000_LED_n */` : This line can make the LEDs light up when the connection rate is 100M.

dengqf6 commented:
Hi, original author here.

> It needs this patch to work: x-wrt@7333152 . ar40xx_phy_mmd_write(priv, 0x1f, 7, 0x8076, 0x0670); /* 1000_LED_n */ : This line can make the LEDs light up when the connection rate is 100M.

The new QCA8075 PHY driver has been merged. To fix the PHY LEDs, you need to add the following nodes to the DTS:
```
&ethphy0 {
	qcom,single-led-1000;
	qcom,single-led-100;
	qcom,single-led-10;
};

&ethphy1 {
	qcom,single-led-1000;
	qcom,single-led-100;
	qcom,single-led-10;
};

&ethphy2 {
	qcom,single-led-1000;
	qcom,single-led-100;
	qcom,single-led-10;
};

&ethphy3 {
	qcom,single-led-1000;
	qcom,single-led-100;
	qcom,single-led-10;
};

&ethphy4 {
	qcom,single-led-1000;
	qcom,single-led-100;
	qcom,single-led-10;
};
```

AmadeusGhost commented:
> The new QCA8075 PHY driver has been merged. To fix the PHY LEDs, you need to add the following nodes to the DTS:

Very thankful.

ysc3839 commented:
@dengqf6 The WAN led is working now. But LAN leds are still not working. Did I missed anything?

AmadeusGhost commented:
> @dengqf6 The WAN led is working now. But LAN leds are still not working. Did I missed anything?

Try https://github.com/openwrt/openwrt/pull/2405

dengqf6 commented:
@robimarko Could you please take a look? All the ethernet LEDs are connected to `1000_LED_n`

@AmadeusGhost That won't work

AmadeusGhost commented:
> All the ethernet LEDs are connected to `1000_LED_n`

It means that we still need https://github.com/x-wrt/x-wrt/commit/58fec253f58c55a2a1dc6e5e5f5f02ad58fe2002 to work?

robimarko commented:
I compared the write you linked to what the PHY driver does now.
There are some differences as PHY driver, will set bits 10, 9, 6, 5, and 4 while the write you linked also sets bits 1 and 0.
So, the only difference should be that LED will blink on traffic while in 10/100 modes but will light up when there is no traffic while in the linked write it will cause LEDs to blink even on just a link and no traffic.

dengqf6 commented:
@robimarko 
It's a broadcast write `0x0670` to register `0x8076`, bit 0 or 1 is not set.

robimarko commented:
@dengqf6 Yeah, sorry.
I completely miscalculated the bits, 0x0670 is exactly what the PHY driver does.
It sets the same bits.

hvkish commented:
Hello, is there any other blockers on this PR? Anything that device owners can help with testing?

adschm commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -682,6 +682,32 @@ define Device/openmesh_a62
+	$(call Device/p2w_r619ac)
+	DEVICE_VARIANT := 64M NAND
+	IMAGES += nand-factory.bin
+	IMAGE/nand-factory.bin := append-ubi | qsdk-ipq-factory-nand
```
While only for 64m? If there a difference, it's not covered in the commit message.

hvkish replied:
Correct me if I'm wrong but I think the modified bootloader can flash normal sysupgrade files, so there is no need to generate a factory binary for that.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-r619ac-128m.dts:
```diff
@@ -0,0 +1,29 @@
+
+			partition@0 {
+				label = "rootfs";
+				reg = <0x0 0x8000000>;
```
Since it's the only effective difference, I'd probably only have put the reg property here (or the partition), and left the rest in the DTSI with a short comment about the missing reg value/partition ... Same for chosen.

leewp14 commented:
@ysc3839 are you still working on this or we create a new PR with the required changes?

ysc3839 commented:
@lee.wp14 I still working on this device. But I have no access to my device at this time. Maybe some weeks later I can test it on my device.

ysc3839 commented:
@adschm Updated, please review.

ysc3839 commented:
I just tested latest commit on my device. The LAN LEDs work only in 1Gbps speed.

iperf3 tests:
5GHz Upload:
```
$ iperf3 -c 192.168.1.1
Connecting to host 192.168.1.1, port 5201
[  5] local 192.168.1.230 port 9181 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.01   sec  30.5 MBytes   253 Mbits/sec
[  5]   1.01-2.00   sec  20.2 MBytes   171 Mbits/sec
[  5]   2.00-3.00   sec  22.6 MBytes   190 Mbits/sec
[  5]   3.00-4.00   sec  19.1 MBytes   160 Mbits/sec
[  5]   4.00-5.00   sec  18.1 MBytes   152 Mbits/sec
[  5]   5.00-6.00   sec  22.6 MBytes   190 Mbits/sec
[  5]   6.00-7.00   sec  23.6 MBytes   198 Mbits/sec
[  5]   7.00-8.00   sec  24.9 MBytes   209 Mbits/sec
[  5]   8.00-9.00   sec  21.9 MBytes   184 Mbits/sec
[  5]   9.00-10.01  sec  21.8 MBytes   181 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.01  sec   225 MBytes   189 Mbits/sec                  sender
[  5]   0.00-10.04  sec   225 MBytes   188 Mbits/sec                  receiver
```
5GHz Download:
```
$ iperf3 -c 192.168.1.1 -R
Connecting to host 192.168.1.1, port 5201
Reverse mode, remote host 192.168.1.1 is sending
[  5] local 192.168.1.230 port 9189 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  35.3 MBytes   296 Mbits/sec
[  5]   1.00-2.00   sec  35.4 MBytes   296 Mbits/sec
[  5]   2.00-3.00   sec  39.1 MBytes   329 Mbits/sec
[  5]   3.00-4.00   sec  45.7 MBytes   383 Mbits/sec
[  5]   4.00-5.00   sec  46.0 MBytes   385 Mbits/sec
[  5]   5.00-6.00   sec  46.1 MBytes   387 Mbits/sec
[  5]   6.00-7.00   sec  45.8 MBytes   384 Mbits/sec
[  5]   7.00-8.00   sec  46.3 MBytes   389 Mbits/sec
[  5]   8.00-9.00   sec  45.2 MBytes   379 Mbits/sec
[  5]   9.00-10.00  sec  46.1 MBytes   386 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.05  sec   431 MBytes   360 Mbits/sec    0             sender
[  5]   0.00-10.00  sec   431 MBytes   361 Mbits/sec                  receiver
```
2.4GHz Upload:
```
$ iperf3 -c 192.168.1.1
Connecting to host 192.168.1.1, port 5201
[  5] local 192.168.1.220 port 10833 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  9.88 MBytes  82.8 Mbits/sec
[  5]   1.00-2.00   sec  9.00 MBytes  75.5 Mbits/sec
[  5]   2.00-3.00   sec  9.25 MBytes  77.6 Mbits/sec
[  5]   3.00-4.00   sec  9.75 MBytes  81.9 Mbits/sec
[  5]   4.00-5.00   sec  10.0 MBytes  83.9 Mbits/sec
[  5]   5.00-6.00   sec  9.25 MBytes  77.6 Mbits/sec
[  5]   6.00-7.00   sec  9.75 MBytes  81.8 Mbits/sec
[  5]   7.00-8.01   sec  10.0 MBytes  83.1 Mbits/sec
[  5]   8.01-9.00   sec  10.2 MBytes  86.7 Mbits/sec
[  5]   9.00-10.01  sec  8.62 MBytes  71.5 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.01  sec  95.8 MBytes  80.2 Mbits/sec                  sender
[  5]   0.00-10.06  sec  95.8 MBytes  79.9 Mbits/sec                  receiver
```
2.4GHz Download:
```
$ iperf3 -c 192.168.1.1 -R
Connecting to host 192.168.1.1, port 5201
Reverse mode, remote host 192.168.1.1 is sending
[  5] local 192.168.1.220 port 10835 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  8.29 MBytes  69.5 Mbits/sec
[  5]   1.00-2.00   sec  9.59 MBytes  80.5 Mbits/sec
[  5]   2.00-3.00   sec  9.99 MBytes  83.8 Mbits/sec
[  5]   3.00-4.00   sec  9.94 MBytes  83.4 Mbits/sec
[  5]   4.00-5.00   sec  9.50 MBytes  79.5 Mbits/sec
[  5]   5.00-6.00   sec  8.51 MBytes  71.5 Mbits/sec
[  5]   6.00-7.00   sec  8.08 MBytes  67.8 Mbits/sec
[  5]   7.00-8.00   sec  6.36 MBytes  53.4 Mbits/sec
[  5]   8.00-9.00   sec  8.27 MBytes  69.3 Mbits/sec
[  5]   9.00-10.00  sec  8.14 MBytes  68.3 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.05  sec  87.0 MBytes  72.7 Mbits/sec    0             sender
[  5]   0.00-10.00  sec  86.7 MBytes  72.7 Mbits/sec                  receiver
```

ygleg commented:
Tested and device works as expected. One nitpick is the non-existent LED in the web interface should be removed:
![image](https://user-images.githubusercontent.com/80977478/114518184-14a81600-9c0d-11eb-8f28-9e12d4d2348f.png)
@ysc3839  Please look into the LED situation.
@adschm This is no longer a WIP and should be ready for use. Please review.

ygleg commented:
## An addendum to the existing flashing instructions:

The _bootloader_ is not what was restricting the device from utilizing the later part of the NAND, but the _partition table_ (0:MIBIB) was the real culprit. MIBIB is read by SBL1 and this information is then passed onto u-boot. I found someone with access to the proprietary QSDK tools that could help me regenerate a new partition table, and I've confirmed that the generated partition table works with my device. The stock bootloader (`U-Boot 2012.07 [Chaos Calmer 15.05.1,unknown] (Dec 27 2017 - 12:06:09)`) does not seem to accept arbitrary keyboard input for interrupting the boot process, however, if for any reason the boot process has failed, it will drop back to a fully functional u-boot shell.

Since everything important/useful is held in the NOR flash, the user may simply erase the entire NAND to get a u-boot shell, and they can update the MIBIB partition in the u-boot shell.

I know the openwrt project usually refrains from bootloader modifications but is typically fine with editing u-boot environment variables, and imo this is similar to that and should be acceptable. If you do a binary diff with the factory MIBIB partition, you'll find only 3 differences, two places have the partition size definition doubled and the checksum is also different. 

Download: https://0x0.st/-Tlu.mibib

### Here's the stock bootloader failing to boot from the 128MB UBI block:
```
U-Boot 2012.07 [Chaos Calmer 15.05.1,unknown] (Dec 27 2017 - 12:06:09)

smem ram ptable found: ver: 1 len: 3
DRAM:  512 MiB
machid : 0x8010006
NAND:  ID = 1d00f101
Vendor = 1
Device = f1
ONFI device found
Is a spansion device needing reset logic
SF: Detected W25Q128 with page size 64 KiB, total 16 MiB
ipq_spi: page_size: 0x100, sector_size: 0x10000, size: 0x1000000
144 MiB
MMC:   qca_mmc: 0
In:    serial
Out:   serial
Err:   serial
machid: 8010006
flash_type: 0
Hit any key to stop autoboot:  0 
Creating 1 MTD partitions on "nand0":
0x000000000000-0x000004000000 : "mtd=0"
UBI: attaching mtd2 to ubi0
UBI: physical eraseblock size:   131072 bytes (128 KiB)
UBI: logical eraseblock size:    126976 bytes
UBI: smallest flash I/O unit:    2048
UBI: VID header offset:          2048 (aligned 2048)
UBI: data offset:                4096
UBI error: vtbl_check: volume table check failed: record 2, error 9
UBI error: ubi_init: cannot attach mtd2
UBI error: ubi_init: UBI error: cannot initialize UBI, error -22
UBI init error 22
qca_mmc: 0
(IPQ40xx) # 
```

### Here's the stock bootlaoder but with the partition table replaced, now able to boot from the 128MB UBI block:
```
U-Boot 2012.07 [Chaos Calmer 15.05.1,unknown] (Dec 27 2017 - 12:06:09)

smem ram ptable found: ver: 1 len: 3
DRAM:  512 MiB
machid : 0x8010006
NAND:  ID = 1d00f101
Vendor = 1
Device = f1
ONFI device found
Is a spansion device needing reset logic
SF: Detected W25Q128 with page size 64 KiB, total 16 MiB
ipq_spi: page_size: 0x100, sector_size: 0x10000, size: 0x1000000
144 MiB
MMC:   qca_mmc: 0
In:    serial
Out:   serial
Err:   serial
machid: 8010006
flash_type: 0
Hit any key to stop autoboot:  0 
Creating 1 MTD partitions on "nand0":
0x000000000000-0x000008000000 : "mtd=0"
UBI: attaching mtd2 to ubi0
UBI: physical eraseblock size:   131072 bytes (128 KiB)
UBI: logical eraseblock size:    126976 bytes
UBI: smallest flash I/O unit:    2048
UBI: VID header offset:          2048 (aligned 2048)
UBI: data offset:                4096
UBI: attached mtd2 to ubi0
UBI: MTD device name:            "mtd=0"
UBI: MTD device size:            128 MiB
UBI: number of good PEBs:        1023
UBI: number of bad PEBs:         1
UBI: max. allowed volumes:       128
UBI: wear-leveling threshold:    4096
UBI: number of internal volumes: 1
UBI: number of user volumes:     3
UBI: available PEBs:             9
UBI: total number of reserved PEBs: 1014
UBI: number of PEBs reserved for bad PEB handling: 10
UBI: max/mean erase counter: 2/1
Read 0 bytes from volume kernel to 84000000
No size specified -> Using max size (2793472)
## Booting kernel from FIT Image at 84000000 ...
   Using 'config@10' configuration
   Trying 'kernel-1' kernel subimage
     Description:  ARM OpenWrt Linux-5.4.105
     Type:         Kernel Image
     Compression:  uncompressed
     Data Start:   0x840000e4
     Data Size:    2699704 Bytes = 2.6 MiB
     Architecture: ARM
     OS:           Linux
     Load Address: 0x80208000
     Entry Point:  0x80208000
     Hash algo:    crc32
     Hash value:   91bd9f40
     Hash algo:    sha1
     Hash value:   6ebbe0b07fac3fad954ba02e32d7af01ccdcd86b
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Flattened Device Tree from FIT Image at 84000000
   Using 'config@10' configuration
   Trying 'fdt-1' FDT blob subimage
     Description:  ARM OpenWrt p2w_r619ac-128m device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x842933d8
     Data Size:    17125 Bytes = 16.7 KiB
     Architecture: ARM
     Hash algo:    crc32
     Hash value:   1ef62435
     Hash algo:    sha1
     Hash value:   0ccefbaa117815f3638e638bdeb0f1c84008ce73
   Verifying Hash Integrity ... crc32+ sha1+ OK
   Booting using the fdt blob at 0x842933d8
   Loading Kernel Image ... OK
OK
   Loading Device Tree to 86ff8000, end 86fff2e4 ... OK
ipq: fdt fixup unable to find compatible node
Using machid 0x8010006 from environment

Starting kernel ...
```

### `smeminfo`
#### Old:
```
(IPQ40xx) # smeminfo
flash_type:             0x6
flash_index:            0x0
flash_chip_select:      0x0
flash_block_size:       0x10000
flash_density:          0x1000000
partition table offset  0x0
No.: Name             Attributes            Start             Size
  0: 0:SBL1           0x0000ffff              0x0          0x40000
  1: 0:MIBIB          0x001040ff          0x40000          0x20000
  2: 0:QSEE           0x0000ffff          0x60000          0x60000
  3: 0:CDT            0x0000ffff          0xc0000          0x10000
  4: 0:DDRPARAMS      0x0000ffff          0xd0000          0x10000
  5: 0:APPSBLENV      0x0000ffff          0xe0000          0x10000
  6: 0:APPSBL         0x0000ffff          0xf0000          0x80000
  7: 0:ART            0x0000ffff         0x170000          0x10000
  8: rootfs           0x0100ffff              0x0        0x4000000
```
#### New:
```
(IPQ40xx) # smeminfo
flash_type:             0x6
flash_index:            0x0
flash_chip_select:      0x0
flash_block_size:       0x10000
flash_density:          0x1000000
partition table offset  0x0
No.: Name             Attributes            Start             Size
  0: 0:SBL1           0x0000ffff              0x0          0x40000
  1: 0:MIBIB          0x001040ff          0x40000          0x20000
  2: 0:QSEE           0x0000ffff          0x60000          0x60000
  3: 0:CDT            0x0000ffff          0xc0000          0x10000
  4: 0:DDRPARAMS      0x0000ffff          0xd0000          0x10000
  5: 0:APPSBLENV      0x0000ffff          0xe0000          0x10000
  6: 0:APPSBL         0x0000ffff          0xf0000          0x80000
  7: 0:ART            0x0000ffff         0x170000          0x10000
  8: rootfs           0x0100ffff              0x0        0x8000000
```

ygleg commented:
## Tested full flashing instructions, MIBIB included

This is done with the factory u-boot, of course you may flash this partition any way you want, including but not limited to: writing this in openwrt after unlocking the mtd block, or pulling off the flash and overwriting it offline. The download link to a working copy is the same as the one attached in my previous comment. https://0x0.st/-Tlu.mibib

1.  Erase the entire NAND in the factory firmware, or simply delete the "kernel" ubi volume with `ubirmvol`. Connect serial console and you will be able to get a u-boot shell after u-boot fails to find the "kernel" volume:
```
Hit any key to stop autoboot:  0 
Creating 1 MTD partitions on "nand0":
0x000000000000-0x000004000000 : "mtd=0"
UBI: attaching mtd2 to ubi0
UBI: physical eraseblock size:   131072 bytes (128 KiB)
UBI: logical eraseblock size:    126976 bytes
UBI: smallest flash I/O unit:    2048
UBI: VID header offset:          2048 (aligned 2048)
UBI: data offset:                4096
UBI: attached mtd2 to ubi0
UBI: MTD device name:            "mtd=0"
UBI: MTD device size:            64 MiB
UBI: number of good PEBs:        512
UBI: number of bad PEBs:         0
UBI: max. allowed volumes:       128
UBI: wear-leveling threshold:    4096
UBI: number of internal volumes: 1
UBI: number of user volumes:     0
UBI: available PEBs:             503
UBI: total number of reserved PEBs: 9
UBI: number of PEBs reserved for bad PEB handling: 5
UBI: max/mean erase counter: 1/1
Volume kernel not found!
qca_mmc: 0
(IPQ40xx) #
```

2. Run `nand info` and make sure your output correlates to what I have here. If it does not, stop immediately and post a comment here. `nand0` (the default) is our real Spansion 128MB NAND, and `nand2` is a fake NAND abstracted from the SPI NOR flash. In this case we do not use the `sf` command for NOR operations, but will use the `nand` command to communicate with the abstracted `nand2` instead.
```
(IPQ40xx) # nand info

Device 0: nand0, sector size 128 KiB
  Page size      2048 b
  OOB size         64 b
  Erase size   131072 b
Device 2: 1090486400x nand2, sector size 64 KiB
  Page size       256 b
  OOB size          0 b
  Erase size    65536 b
```

3. Switch the active device to the SPI flash
```
(IPQ40xx) # nand device 2
Device 2: nand2... is now current device
```

4. Launch a TFTP server at 192.168.1.10 on your computer and fetch the new MIBIB into memory
```
(IPQ40xx) # tftpboot 0x84000000 r619ac-16m-128m.mibib
MAC0 addr: xx:xx:xx:xx:xx:xx
PHY ID1: 0x4d
PHY ID2: 0xd0b1
ipq40xx_ess_sw_init done
eth0
eth0 PHY0 Down Speed :10 Half duplex
eth0 PHY1 Down Speed :10 Half duplex
eth0 PHY2 Down Speed :10 Half duplex
eth0 PHY3 up Speed :1000 Full duplex
eth0 PHY4 Down Speed :10 Half duplex
Using eth0 device
TFTP from server 192.168.1.10; our IP address is 192.168.1.11
Filename 'r619ac-16m-128m.mibib'.
Load address: 0x84000000
Loading: T T T #########
done
Bytes transferred = 131072 (20000 hex)
```

5. Erase the old MIBIB partition.
Note: `nand erase.part` seems to be broken on the factory u-boot, otherwise I would have used that instead.
```
(IPQ40xx) # nand erase 0x40000 0x20000

NAND erase: device 2 offset 0x40000, size 0x20000
Erasing at 0x50000 -- 100% complete.
OK
```

6. Write the new partition table to flash
```
(IPQ40xx) # nand write 0x84000000 0x40000 0x20000

NAND write: device 2 offset 0x40000, size 0x20000
 131072 bytes written: OK
```
### Installing OpenWrt
7. Switch back to `nand0` 
**THIS IS VERY IMPORTANT, MAKE SURE YOU DID THIS BEFORE PROCEEDING**
```
(IPQ40xx) # nand device 0
Device 0: nand0... is now current device
```

8. Erase the NAND
```
(IPQ40xx) # nand erase.chip

NAND erase.chip: device 0 whole chip
Erasing at 0x7fe0000 -- 100% complete.
OK
```

9. Fetch the compiled UBI blob over TFTP to memory and write it to NAND. Note that the size 0x680000 may change, use whetever tftp prints instead of blindly copying the command. `nand write 0x84000000 0x0 $SIZE_HERE`
```
(IPQ40xx) # tftp 0x84000000 openwrt-ipq40xx-generic-p2w_r619ac-128m-squashfs-nand-factory.ubi
eth0 PHY0 Down Speed :10 Half duplex
eth0 PHY1 Down Speed :10 Half duplex
eth0 PHY2 Down Speed :10 Half duplex
eth0 PHY3 up Speed :1000 Full duplex
eth0 PHY4 Down Speed :10 Half duplex
Using eth0 device
TFTP from server 192.168.1.10; our IP address is 192.168.1.11
Filename 'openwrt-ipq40xx-generic-p2w_r619ac-128m-squashfs-nand-factory.ubi'.
Load address: 0x84000000
Loading: T #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         ##########
done
Bytes transferred = 6815744 (680000 hex)
(IPQ40xx) # nand write 0x84000000 0x0 0x680000

NAND write: device 0 offset 0x0, size 0x680000
 6815744 bytes written: OK
```

10. Reboot
```
(IPQ40xx) # reset
resetting ...
```

Now you should see the device start up and working as expected, with the full 128MB available:
```
root@OpenWrt:/# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 3.5M      3.5M         0 100% /rom
tmpfs                   248.2M     56.0K    248.1M   0% /tmp
/dev/ubi0_2             105.6M     48.0K    100.8M   0% /overlay
overlayfs:/overlay      105.6M     48.0K    100.8M   0% /
tmpfs                   512.0K         0    512.0K   0% /dev
```

ygleg commented:
@chunkeey Would you mind taking a look at this and see if there are any loose ends?

chunkeey commented:
Oh boy. You need to have a "Signed-off-by: " or something equivalent from @LGA1150 for this.

Also, please follow the comment in ipq-wifi and post your board file changes upstream:
https://github.com/openwrt/openwrt/blob/master/package/firmware/ipq-wifi/Makefile#L23

(And then there's bootargs-append. I don't know what arguments are passed from the bootloader, but squashfs is already the standard. The same goes for root=.... Suggestion: Have you tried to rename the "rootfs" label to "ubi"? Because this should also help with sysupgrade from one OpenWrt to the next. Again, that "rootfs" label for the ubi-partition looks to break the current sysupgrade code. This is because the default CI_UBIPART is "ubi" and not "rootfs". "rootfs" is the CI_ROOTPART volume which is inside the ubi-partition.)

ysc3839 commented:
@chunkeey Is there any document about the `label` value in dts? I don't know what's the difference between `rootfs` and `ubi`.

ysc3839 commented:
@chunkeey After removing `bootargs-append` the kernel unable to mount rootfs.
Here is the kernel log from serial: [klog.txt](https://github.com/openwrt/openwrt/files/6373186/klog.txt)
The bootloader passed `ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs rootwait` to kernel.

ysc3839 commented:
@chunkeey
> Oh boy. You need to have a "Signed-off-by: " or something equivalent from @LGA1150 for this.

Do you mean I have to ask @LGA1150 for a "Signed-off-by: " in order to allow this pr merged?

> Also, please follow the comment in ipq-wifi and post your board file changes upstream:

Is this required for all devices? There are other deivces not submitted board file to upstream: https://github.com/openwrt/openwrt/tree/master/package/firmware/ipq-wifi

LGA1150 commented:
> Do you mean I have to ask @LGA1150 for a "Signed-off-by: " in order to allow this pr merged?

Signed-off-by: DENG Qingfang ***@***.***>

imi415 commented:
@ysc3839 The mini-PCIe power supply is controlled by a MOSFET which requires a hardware mod to override, however, it turns out that the gate of the MOSFET is controlled by GPIO2, we could just add a dedicated pin in dt `&tlmm` node to forcelly enable that output, e.g.
```
+       pcie_pins: pcie_pinmux {
+               mux {
+                       pins = "gpio2";
+                       function = "gpio";
+                       bias-pull-down;
+                       output-low;
+               };
+       };
```
What do you think?

imi415 commented:
If I get this correctly, @chunkeey ‘s point is if modification to stock kernel parameters  is required, then `bootargs-append` property should be kept at minimal. E.g. no duplicated `rootfstype=squashfs` should be listed. So here's the minimal bootable `bootargs-append` entry below.
```patch
 / {
        model = "P&W R619AC 128M";
        compatible = "p2w,r619ac-128m";
+
+       chosen {
+               bootargs-append = " ubi.mtd=ubi root=/dev/ubiblock0_1"; //Keep the leading space!
+       };
 };

```

This works with NAND mtd label set to `ubi`, and rootfs from ubi.

ysc3839 commented:
I just resolved conflicts. Now I'm compiling and testing.

ysc3839 commented:
After updated to latest master, 802.11ac become slow.
LuCI:
```
Qualcomm Atheros IPQ4019 802.11nac
Channel: 149 (5.745 GHz) | Bitrate: 54 Mbit/s
```
iperf3:
```
$ iperf3 -c 192.168.1.1 -R
Connecting to host 192.168.1.1, port 5201
Reverse mode, remote host 192.168.1.1 is sending
...
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.01  sec  32.2 MBytes  26.9 Mbits/sec    0             sender
[  5]   0.00-10.00  sec  31.8 MBytes  26.7 Mbits/sec                  receiver
```
I will test 21.02 branch.

ayyyuki1 commented:
I registered on github just to say thanks for this, I now have my R619AC / G-DOCK 2.0 fully working without using those kind of different chinese-GFW-internet-penetration forks. Its a good device. 

As for issues, the only Issue I see compared to coolsnowwolf's fork is that I can't seem to use the WAN port for VLANs? but other than that, everything just works ootb, benching 375Mbps on 5GHz wifi on iperf3.

ysc3839 commented:
@ayyyuki1 WAN port is on a separate network device, are you sure it can use VLAN?

ayyyuki1 commented:
> @ayyyuki1 WAN port is on a separate network device, are you sure it can use VLAN?

I cannot use VLAN on this openwrt on the WAN port

ygleg commented:
> Oh boy. You need to have a "Signed-off-by: " or something equivalent from @LGA1150 for this.

I think this is fixed. https://github.com/openwrt/openwrt/pull/3691#issuecomment-826475714

> Also, please follow the comment in ipq-wifi and post your board file changes upstream: https://github.com/openwrt/openwrt/blob/master/package/firmware/ipq-wifi/Makefile#L23

[Boardfiles are submitted to upstream](https://lists.infradead.org/pipermail/ath10k/2021-November/013059.html
). Unfortunately I forgot to include the ath10k-firmware header in the email title so I might have to re-send it at some point. I would really prefer if this can get merged right now without having this issue as a blocker, since there were numerous attempts with getting this particular device added without success, and this PR in particular is in a relatively healthy state imo.

> (And then there's bootargs-append. I don't know what arguments are passed from the bootloader, but squashfs is already the standard. The same goes for root=.... Suggestion: Have you tried to rename the "rootfs" label to "ubi"? Because this should also help with sysupgrade from one OpenWrt to the next. Again, that "rootfs" label for the ubi-partition looks to break the current sysupgrade code. This is because the default CI_UBIPART is "ubi" and not "rootfs". "rootfs" is the CI_ROOTPART volume which is inside the ubi-partition.)

This is resolved I believe. https://github.com/openwrt/openwrt/pull/3691#issuecomment-920156132

> (EDIT: Please take a quick peek at https://openwrt.org/submitting-patches as well . For example, there's a "max. 75-ish characters per line in the commit message" limit. This is because people still use these 80x25 terminals. Of course, if there's a link or a long filename, one-line c&p install instructions, etc... in the commit message that requires more than 75 characters per line, then that is not enforced. So please reformat your commit message accordingly. Thanks. )

@ysc3839 Have you addressed these issues?

@chunkeey Could you please take a look this PR again to see if there are any lingering issues? Thanks.

chunkeey commented:
Ok, I see that you have posted the device's boardfiles to ath10k. :+1: 

http://lists.infradead.org/pipermail/ath10k/2021-November/013059.html

I wonder why @mailfence had to be used? Posting with a business e-mail-address
will usually get things going... Especially because Qualcomm likes to do Business!

ysc3839 committed:
ipq40xx: add support for P&W R619AC (aka G-DOCK 2.0)

P&W R619AC is a IPQ4019 Dual-Band AC1200 router.
It is made by P&W (p2w-tech.com) known as P&W R619AC but marketed and sold
more popularly as G-DOCK 2.0.

Specification:

* SOC: Qualcomm Atheros IPQ4019 (717 MHz)
* RAM: 512 MiB
* Flash: 16 MiB (NOR) + 128 MiB (NAND)
* Ethernet: 5 x 10/100/1000 (4 x LAN, 1 x WAN)
* Wireless:
  - 2.4 GHz b/g/n Qualcomm Atheros IPQ4019
  - 5 GHz a/n/ac Qualcomm Atheros IPQ4019
* USB: 1 x USB 3.0
* LED: 4 x LAN, 1 x WAN, 2 x WiFi, 1 x Power (All Blue LED)
* Input: 1 x reset
* 1 x MicroSD card slot
* Serial console: 115200bps, pinheader J2 on PCB
* Power: DC 12V 2A
* 1 x Unpopulated mPCIe Slot
* 1 x Unpopulated Sim Card Slot

Installation:
1. Access to tty console via UART serial
2. Enter failsafe mode and mount rootfs
   (https://openwrt.org/docs/guide-user/troubleshooting/failsafe_and_factory_reset)
3. Edit inittab to enable shell on tty console
   `sed -i 's/#ttyM/ttyM/' /etc/inittab`
4. Reboot and upload `-nand-factory.bin` to the router (using wget)
5. Use `sysupgrade` command to install

Another installation method is to hijack the upgrade server domain of stock
firmware, because it's using insecure http.

This commit is based on @LGA1150(at GitHub)'s work
(https://github.com/LGA1150/openwrt/commit/a4932c8d5a275d1fb4297bd20ec03f9270a45d1c)
With some changes:
1. Added `qpic_bam` node in dts. I don't know much about this,
   but I observed other dtses have this node.
2. Removed `ldo` node under `sd_0_pinmux`, because `ldo` cause SD card not
   working. This fix is from
   https://github.com/coolsnowwolf/lede/commit/51143b4c7571f717afe071db60bbb4db1532cbf2
3. Removed the 32MB NOR variant.
4. Removed `cd-gpios` in `sdhci` node, because it's reported that it makes
   wlan2g led light up.
5. Added ethphy led config in dts.
6. Changed nand partition label from `rootfs` to `ubi`.

About the 128MiB variant: The stock bootloader sets size of nand to 64MiB.
But most of this devices have 128MiB nand. If you want to use all 128MiB,
you need to modify the `MIBIB` data of bootloader.

Signed-off-by: Richard Yu <yurichard3839@gmail.com>
Signed-off-by: DENG Qingfang <dqfext@gmail.com>

ysc3839 commented:
@ygleg Yes, the commit message is edited not exceed 75 chars per line.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-r619ac.dtsi:
```diff
@@ -0,0 +1,340 @@
+
+/ {
+	chosen {
+		bootargs-append = " ubi.mtd=ubi root=/dev/ubiblock0_1";
```
hmm, the bootargs-append should not be needed anymore. 
This is because the partition label "ubi" will trigger OpenWrt UBIs autoloader to do the right thing.

(Still, you don't need to remove them.)

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-r619ac.dtsi:
```diff
@@ -0,0 +1,340 @@
+};
+
+&ethphy0 {
+	qcom,single-led-1000;
```
So, sadly these bindings have been rejected upstream...

https://www.spinics.net/lists/devicetree/msg397612.html

But I see @adschm has merged the qca807x.c which can do something with them.
So these are now openwrt specific. If Qualcomm ever bothers and upstreams a 
qca807x driver, then this all needs to be updated.

Oh well. (No need to do anything here)

adschm replied:
Did I?

chunkeey replied:
> Did I?

nope, that's my bad. Sorry.

chunkeey commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -753,6 +753,32 @@ define Device/openmesh_a62
+
+define Device/p2w_r619ac-128m
+	$(call Device/p2w_r619ac)
+	DEVICE_VARIANT := 128M NAND
```
With your comment from the commit message in mind:

> About the 128MiB variant: The stock bootloader sets size of nand to 64MiB.
But most of this devices have 128MiB nand. If you want to use all 128MiB,
you need to modify the `MIBIB` data of bootloader.

Hmm, can you provide instructions how to modifiy the MIBIB?

Can the bootloader read a kernel from ubi, if it has blocks placed at and over the 64MiB mark?
(Reason that this can happen is when UBI has to work around a bad block and remaps it "somewhere else",
so having something that just needs to be copy&pasted into a shell as part of the install instructions
would be helpful)

ysc3839 replied:
It's provided by @ygleg [here](https://github.com/openwrt/openwrt/pull/3691#issuecomment-818770060). I have not tested this because I don't have tools to flash the bootloader chip. Now I'm using a third-party bootloader.

imi415 replied:
I have tested the MIBIB file attached above, which works as intended.
It seems that NOR flash MTDs can not be recognized by `mtd` utility, making the flash procedure only viable (without an external SPI flash programmer) by erasing the full NAND, which will make stock u-boot fallback to a working shell. After that, the partition could be updated by using integrated TFTP and mtd command.
There are also 3rd-party bootloaders which support the new (same? since the only partition on NAND is the rootfs ubi) 128MB partition scheme.

chunkeey replied:
Yea, I this would be something for the wiki. For now, could you add a link to the github in the commit message? This is to point people into the right direction and to that helpful comment.

ygleg replied:
You may also short any two of the data pins together on the NAND with a tweezer (pinout can be found in the ONFI specification or the datasheet, it's standard), that will cause u-boot to fail while loading the kernel and drop into a shell.

chunkeey commented on target/linux/ipq40xx/patches-5.4/901-arm-boot-add-dts-files.patch:
```diff
@@ -58,6 +58,8 @@ Signed-off-by: John Crispin <john@phrozen.org>
 +	qcom-ipq4019-e2600ac-c2.dtb \
 +	qcom-ipq4019-habanero-dvk.dtb \
 +	qcom-ipq4019-pa2200.dtb \
++	qcom-ipq4019-r619ac-64m.dtb \
```
While the 5.4 patches are sticking around, ipq40xx no longer has 5.4 marked as a supported kernel.
Can you please update this to 5.10?

ysc3839 replied:
Updated.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-r619ac-128m.dts:
```diff
@@ -0,0 +1,12 @@
+};
+
+&nand_rootfs {
+	reg = <0x0 0x8000000>;
```
A comment about the bootloader's 64MiB limit would be very thoughtful. 

(At this point, I prefer if the whole nand node was copied, even if just the size in the reg property is different.
But that's just me, you don't need to change it.)

ygleg replied:
Seconded, I agree that there should be a comment mentioning the stock MIBIB being set up for a 64MB chip,

ygleg commented:
Just so we have all the useful hardware info in one place and someone can take full advantage of this device without fumbling around on the internet, here are some more notes on the hardware itself:

The REFCLK signals on the Mini PCIe slot is not connected on this board ootb. If you want to use the Mini PCIe slot on the board, you need to (preferably) solder two 0402 resistors: **R436** (REFCLK+) and **R444** (REFCLK-). I don't have the BOM, so I don't know what exact values those are supposed to be, but an educated guess would be 22Ω. I put these on my board and u-boot can correctly detect PCIe devices now, but I haven't tested this extensively. Supposedly you can also short those with some solder bridges, but I have no idea how good of a solution that is. These resistors can be located at the bottom side of the board, right under where the MiniPCIe connector is at.

Additionally, if your Mini PCIe wireless NIC needs auxiliary power input on the non-standard pins for the RF frontend (check with your vendor), you can also add 0603 0Ω jumpers to enable this functionality. Bridging **R443** will supply power to pin **45** and **47**, and bridging **R447** will supply power to pin **49** and **51**. Do not bridge these unless you know what you are doing as these are non-standard things that can potentially fry other Mini PCIe devices. These two jumpers can also be located at the bottom side of the board, right under where the Mini PCIe connector is at.

The 3v3 power supply on the Mini PCIe slot is controlled by a MOSFET attached to gpio2 (see dts on this pull request), previously a 1206 0Ω jumper needs to be added to **R489** (right by the MOSFET) to bypass the MOSFET, but it shouldn't be required anymore now that we have the MOSFET turned on by default.

chunkeey commented:
Merged. Thanks! (I added some notes here and there. Hope they are useful and left the other stuff as is.)

ygleg commented:
Thanks to everyone who's involved in this PR, I'm really glad to see this finally getting merged after 11 months!

ysc3839 commented:
@chunkeey Can I backport this device to 21.02?

chunkeey commented:
You are talking about "minew g1-c" and "iEi Puzzle-M901/Puzzle-M902" which were added after 21.02(.0)?
I don't know, if it was intended. From what I think I "knew" the openwrt release branch is there for fixes.
But I can ask at the next meeting.

In the meantime, since you seem to be a fountain of knowledge, it would be great to have some of the posts content also in the wiki as well. If you feel like writing this down: you can request a wiki account through the forum:
https://forum.openwrt.org/t/applying-for-openwrt-wiki-account/101671

Cheers
Christian

And yes: I did mess it up with the missing include. Sorry for that, because otherwise we probably would have a snapshot image for the device by now. Please test those too! This is because the buildbot are generating the images slightly differently and something new could turn up there, that wasn't considered before.

ysc3839 commented:
I'm not talking about those two devices. Just want to know if it's allowed to do.

ygleg commented:
chunkeey was referring to the fact that new devices are backported to 21.02 despite the [wiki page on device backporting policy](https://openwrt.org/docs/guide-developer/device-support-policies#device_support_backports) saying it's unlikely to happen. This pull request was initially meant for kernel 5.4, which is what 21.04 is on for ipq40xx as well, so as long as this doesn't require any new drivers in 5.10 that 5.4 does not have, it should be a "trivial" change. With that said, we're getting into some murky waters on this because "trivial changes" are sometimes accepted and sometimes rejected:
- minew g1-c (accepted): https://github.com/openwrt/openwrt/commit/7f7bf36ec5bacbf414615d3cde009be548a86ae0
- iEi Puzzle-M901/Puzzle-M902 (accepted): https://github.com/openwrt/openwrt/commit/c43a5921fa7288ba183cc56da8f110a6ed0cd958
- Xiaomi AIoT Router AC2350 (rejected): https://github.com/openwrt/openwrt/pull/4630

From what I can tell, it's very unlikely that a cherry-pick PR on github is going to be accepted, but if @chunkeey is willing to cherry-pick it and just commit it, we might have a chance at this, since the changed has just been reviewed and it's unlikely going to cause any regressions. It would indeed be very nice if this device can be backported to 21.02.
